casper/src/main/scala/coop/rchain/casper/blocks/merger/CasperDagMerger.scala

```diff
@@ -2,83 +2,214 @@ package coop.rchain.casper.blocks.merger
 
 import cats.effect.Concurrent
 import cats.syntax.all._
+import com.google.protobuf.ByteString
 import coop.rchain.casper.PrettyPrinter
 import coop.rchain.casper.protocol.ProcessedDeploy
 import coop.rchain.casper.util.EventConverter
 import coop.rchain.casper.util.rholang.RuntimeManager
 import coop.rchain.casper.util.rholang.RuntimeManager.StateHash
 import coop.rchain.crypto.codec.Base16
 import coop.rchain.dag._
+import coop.rchain.metrics.Metrics.Source
+import coop.rchain.metrics.{Metrics, Span}
 import coop.rchain.rspace.Blake2b256Hash
 import coop.rchain.rspace.merger.EventChain
+import coop.rchain.rspace.merger.instances.EventsIndexConflictDetectors
 import coop.rchain.rspace.syntax._
+import coop.rchain.shared.syntax._
 import coop.rchain.shared.{Log, Stopwatch}
 import coop.rchain.store.LazyKeyValueCache
 
 import scala.collection.Seq
 
+/**
+  * Merger optimizing the process using blocks relation from DAG.
+  * DAG is split into non intersecting branches, which are merged afterwards.
+  */
 object CasperDagMerger {
+  val indexBranchesLabel: Source      = Metrics.Source(MergingMetricsSource, "index-branches")
+  val calculateConflictsLabel: Source = Metrics.Source(MergingMetricsSource, "calculate-conflicts")
+  val mergeBranchesLabel: Source      = Metrics.Source(MergingMetricsSource, "merge-branches")
+
+  /** branch having start state, end state and index for performant merging */
+  final case class IndexedBranch(
+      startState: ByteString,
+      endState: ByteString,
+      index: BranchIndex
+  )
+
+  /** conflict resolution for a sequence of branches */
+  final case class ConflictResolution(
+      mainBranch: IndexedBranch,
+      mergingBranches: Seq[IndexedBranch],
+      channelsConflicting: Set[Blake2b256Hash],
+      deploysToReject: Set[ProcessedDeploy]
+  )
 
   /**
     * Merge DAG.
     * @param tips - top blocks of the DAG
     * @param base - bottom blocks of the DAG, all tips should be connected with base
     * @param dag - connection graph
-    * @return tuplespace state hash of the merged state and list of conflicting deploys to reject
+    * @return tuple space state hash of the merged state and list of conflicting deploys to reject
     */
-  def merge[F[_]: Concurrent: RuntimeManager: Log](
+  def merge[F[_]: Concurrent: RuntimeManager: Log: Span: Metrics](
       tips: Seq[MergingVertex],
       base: MergingVertex,
-      dag: DagReader[F, MergingVertex]
+      dag: DagReader[F, MergingVertex],
+      blockIndexCache: LazyKeyValueCache[F, MergingVertex, BlockIndex]
   ): F[(StateHash, Seq[ProcessedDeploy])] = {
 
-    // ordering is required for computing branches procedure to be replayable
+    // ordering is required to be able to replay computing branches procedure
     implicit val ord: Ordering[MergingVertex] = (x: MergingVertex, y: MergingVertex) =>
       Base16.encode(x.postStateHash.toByteArray).compare(Base16.encode(y.postStateHash.toByteArray))
```

Comment:
Encoding to string can be avoided and at least compare bytes. Right now I don't see better solution, ByteString only have lexical comparison which should be slower.
`java.util.Arrays.compare(x.postStateHash.toByteArray, y.postStateHash.toByteArray)`

Reply:
so we are leaving it as it is?

Reply:
What I suggested is to use array comparison directly without conversion to string and comparing strings, which should be slower.